### PR TITLE
PacketBuffer: Make discontiguous LwIP pbuf adoptable

### LINKED ISSUE #21816: Try to reuse the lwIP pbuf in UDPEndPointImplLwIP.cpp
#### Problem

Currently all the UDP packets are copied to fit the `PacketBuffer` memory model.

 #### Proposed Solution

Check the memory layout of the lwIP pbuf and reuse it if it already fits the `PacketBuffer`.

### PR DESCRIPTION
**Pull request by wqx6**
### Changes
Make the discontiguous LwIP pbuf adoptable to reduce the `PacketBuffer: pool EMPTY.` errors.

### Tests
Tested on ESP32, both the unicast and multicast message can be received and handled well.

### CODE REVIEW

File: src/system/SystemPacketBuffer.cpp [hunk 1]
```diff
@@ -236,6 +241,12 @@ void PacketBuffer::SetDataLength(uint16_t aNewLen, PacketBuffer * aChainHead)
 
 uint16_t PacketBuffer::MaxDataLength() const
 {
+#if CHIP_SYSTEM_CONFIG_USE_LWIP
+    if (!(PBUF_STRUCT_DATA_CONTIGUOUS(this)))
+    {
+        return static_cast<uint16_t>(DataLength());
```
Comment (bzbarsky-apple) [inline src/system/SystemPacketBuffer.cpp:247]: Why is that cast needed?  `DataLength` returns `uint16_t`.

File: src/system/SystemPacketBuffer.cpp [hunk 2]
```diff
@@ -253,11 +264,23 @@ uint16_t PacketBuffer::ReservedSize() const
 
 uint8_t * PacketBuffer::ReserveStart()
 {
+#if CHIP_SYSTEM_CONFIG_USE_LWIP
+    if (!(PBUF_STRUCT_DATA_CONTIGUOUS(this)))
+    {
+        return reinterpret_cast<uint8_t *>(this->payload);
```
Comment (bzbarsky-apple) [inline src/system/SystemPacketBuffer.cpp:270]: Any reason not to just return `Start()` here?

File: src/system/SystemPacketBuffer.cpp [hunk 3]
```diff
@@ -253,11 +264,23 @@ uint16_t PacketBuffer::ReservedSize() const
...
+#endif
     return reinterpret_cast<uint8_t *>(this) + kStructureSize;
 }
 
 const uint8_t * PacketBuffer::ReserveStart() const
 {
+#if CHIP_SYSTEM_CONFIG_USE_LWIP
+    if (!(PBUF_STRUCT_DATA_CONTIGUOUS(this)))
+    {
+        return reinterpret_cast<const uint8_t *>(this->payload);
```
Comment (bzbarsky-apple) [inline src/system/SystemPacketBuffer.cpp:281]: Why not return `Start()`?

**Review by bzbarsky-apple — COMMENTED** [review]
This fixes https://github.com/project-chip/connectedhomeip/issues/21816, right?

It looks to me like the CloneData bits described in https://github.com/project-chip/connectedhomeip/pull/20923#issuecomment-1204501799 are not an issue because CloneData only uses public APIs that this PR fixes, right?

File: src/system/SystemPacketBuffer.cpp (hunk 1 above)
Comment (wqx6) [reply src/system/SystemPacketBuffer.cpp]: Removed the cast

File: src/system/SystemPacketBuffer.cpp (hunk 2 above)
Comment (wqx6) [reply src/system/SystemPacketBuffer.cpp]: Changed

File: src/system/SystemPacketBuffer.cpp (hunk 3 above)
Comment (wqx6) [reply src/system/SystemPacketBuffer.cpp]: Changed

**Comment by wqx6** [comment]
> This fixes #21816, right?
> 
> It looks to me like the CloneData bits described in [#20923 (comment)](https://github.com/project-chip/connectedhomeip/pull/20923#issuecomment-1204501799) are not an issue because CloneData only uses public APIs that this PR fixes, right?

we might run out of PBUF pool if we clone the udp message. After we increase the event queue, the error `PacketBuffer: pool EMPTY` happens when the ScheduleLambda is block and the udp endpoint receives a lot of packets. Described in [#23180 comment](https://github.com/project-chip/connectedhomeip/issues/23180#issuecomment-1278272965).

**Comment by bzbarsky-apple** [comment]
> we might run out of PBUF pool if we clone the udp message.

Yes, I understand that.  My questions are:

1. Does this fix https://github.com/project-chip/connectedhomeip/issues/21816?
2. With these changes, is calling CloneData on a non-contiguous buffer that's been adopted (which some of our internals will end up doing) going to do the right thing?

**Comment by bzbarsky-apple** [comment]
And perhaps most concretely: was receiving group messages tested with this change?

**Comment by wqx6** [comment]
> Does this fix https://github.com/project-chip/connectedhomeip/issues/21816?

Yes

> With these changes, is calling CloneData on a non-contiguous buffer that's been adopted (which some of our internals will end up doing) going to do the right thing?

Yes, we will create a pbuf with the reserve size 0 in `CloneData` and copy the payload. Same as https://github.com/project-chip/connectedhomeip/pull/25679/files#diff-d518c09d6ad62381f954eb9520bb175375404f40576a60fa6b25f37bc1c288c0L423.

> was receiving group messages tested with this change?

Yes, I have tested the group messages receiving on ESP32.

**Review by bzbarsky-apple — APPROVED** [review]
Thanks.  That all matches what I was seeing too!

**Review by andy31415 — APPROVED** [review]

**Pull request merged**